bug: make sure that controllers run on controll plane nodes

vespian opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Fixes #2361


**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. 
2. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE"....however we encourage contributors to never use this as release notes are incredible useful.
-->
```release-note
Cluster API Provider AWS will now be preferentially pinned to control plane nodes. This is especially helpful when running self-managed management clusters in AWS as for EC2-based control planes, the control plane EC2 instances have the controlplane.cluster-api.sigs.k8s.io IAM role which has sufficient permissions for CAPA to run.
Action Required: Please ensure your control plane nodes have sufficient resources to run Cluster API Provider AWS.
```

sedefsavas commented:
Thanks for the PR, we opened the main branch for v1alpha4 development. Can you rebase the PR?

dkoshkin commented on bootstrap/eks/config/manager/manager.yaml:
```diff
@@ -26,3 +26,10 @@ spec:
       tolerations:
         - effect: NoSchedule
           key: node-role.kubernetes.io/master
+      affinity:
+       nodeAffinity:
+         requiredDuringSchedulingIgnoredDuringExecution:
+           nodeSelectorTerms:
+           - matchExpressions:
+             - key: node-role.kubernetes.io/master
+               operator: Exists
```
As `node-role.kubernetes.io/master` label is going away should we take this opportunity to include the new `node-role.kubernetes.io/control-plane` label to support future versions of Kubernetes (while keeping the old label for now)
```suggestion
        - effect: NoSchedule
          key: node-role.kubernetes.io/control-plane
        - effect: NoSchedule
          key: node-role.kubernetes.io/master
      affinity:
       nodeAffinity:
         requiredDuringSchedulingIgnoredDuringExecution:
           nodeSelectorTerms:
           - matchExpressions:
             - key: node-role.kubernetes.io/control-plane
               operator: Exists
           # remove once usage of node-role.kubernetes.io/master is removed from Kubernetes           
           - matchExpressions:           
             - key: node-role.kubernetes.io/master
               operator: Exists
```

vespian commented:
@sedefsavas @dkoshkin All comments were addressed, please have another look.

randomvariable commented:
/hold

We need to consider the different labels across different Kubernetes versions. Will discuss with SIG CL folk upstream first.

randomvariable commented on bootstrap/eks/config/manager/manager.yaml:
```diff
@@ -25,5 +25,18 @@ spec:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: node-role.kubernetes.io/control-plane
+                operator: Exists
+            # remove once usage of node-role.kubernetes.io/master is removed from Kubernetes
+            - matchExpressions:
+              - key: node-role.kubernetes.io/master
+                operator: Exists
```
Given these expressions will be AND'd, then this won't work on older versions of Kubernetes that don't set control-plane, and will also stop working when `master` is removed.

@neolit123 , was there any guidance for appdevs on how best to handle this?

neolit123 replied:
we do not have a guide for node affinity and we assumed that k8s constructs are flexible enough to manage this, are well documented, and users will know how to use them.

you can ask someone to amend the KEP if you think it's worthy:
https://github.com/kubernetes/enhancements/tree/master/keps/sig-cluster-lifecycle/kubeadm/2067-rename-master-label-taint

https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity

> If you specify multiple nodeSelectorTerms associated with nodeAffinity types, then the pod can be scheduled onto a node if one of the nodeSelectorTerms can be satisfied.

> If you specify multiple matchExpressions associated with nodeSelectorTerms, then the pod can be scheduled onto a node only if all matchExpressions is satisfied.

i'm reading this as - one of the nodeSelectorTerms is a sufficient match, but if a nodeSelectorTerms has multiple matchExpressions they must all match within it.

randomvariable replied:
Ah yes, multiple nodeSelectorTerms will make it work. Cool, I think that's fine then.

Thanks for that!

randomvariable commented:
/unhold
/ok-to-test

randomvariable commented:
/override pull-cluster-api-provider-aws-build-release-0-6

randomvariable commented:
We have one issue with this PR. And that is for EKS, there are no control plane instances, so an EKS management cluster isn't going to work here. We need a way to turn this on optionally.

randomvariable commented:
Placing a 
/hold
on this until we can sort out a solution to cover EKS.

sedefsavas commented:
cc @richardcase any idea for how we want to do this in EKS side?

I think it would be nice to backport this to v0.6.x release as well.

richardcase commented:
> cc @richardcase any idea for how we want to do this on EKS side?

Labels/affinity would work. If we wanted to segregate CAPA in EKS we can create machine pools with node labels.

richardcase commented on bootstrap/eks/config/manager/manager.yaml:
```diff
@@ -25,5 +25,18 @@ spec:
           key: node-role.kubernetes.io/master
+      affinity:
+        nodeAffinity:
+          requiredDuringSchedulingIgnoredDuringExecution:
```
Could we change this to `preferredDuringSchedulingIgnoredDuringExecution` instead?

richardcase commented on bootstrap/eks/config/manager/manager.yaml:
```diff
@@ -25,5 +25,18 @@ spec:
       terminationGracePeriodSeconds: 10
       tolerations:
+        - effect: NoSchedule
+          key: node-role.kubernetes.io/control-plane
```
I was also thinking we should make this key configurable like this:

```
key: ${K8S_CP_LABEL:=node-role.kubernetes.io/control-plane} 
```

so that when using EKS you can taint nodes to run "system" services.

richardcase commented on bootstrap/eks/config/manager/manager.yaml:
```diff
@@ -25,5 +25,18 @@ spec:
+          requiredDuringSchedulingIgnoredDuringExecution:
+            nodeSelectorTerms:
+            - matchExpressions:
+              - key: node-role.kubernetes.io/control-plane
```
As above.

richardcase reviewed (changes requested):
@vespian - this is looking good. A few suggestions to help with EKS

vespian commented:
@richardcase I addressed your comments, changes in commit `'make control plane tolerations configurable, change to preferredDuringSchedulingIgnoredDuringExecution'`. 

Please have another look.

sedefsavas commented:
Changes LGTM.
@vespian could you check why this test is failing:
`    --- FAIL: TestEKSConfigReconciler/Should_reconcile_an_EKSConfig_and_update_data_Secret (0.09s)
`

randomvariable commented:
> --- FAIL: TestEKSConfigReconciler/Should_reconcile_an_EKSConfig_and_update_data_Secret (0.09s)

I think it needs rebasing on main. `Should_reconcile_an_EKSConfig_and_update_data_Secret` wasn't testing what it thought it was testing.

randomvariable commented:
/test pull-cluster-api-provider-aws-test

randomvariable commented:
/test pull-cluster-api-provider-aws-e2e-blocking

randomvariable commented:
/unhold
/lgtm
/approve
